functorch/_src/memory_efficient_op_authoring.py

```diff
@@ -0,0 +1,159 @@
+            # By wrapping the arguments with proxies,
+            # we can dispatch to the appropriate
+            # decomposition rule and implicitly add it
+            # to the Graph by symbolically tracing it.
+            tracer = GraphAppendingTracer(new_graph)
+            proxy_args = [
+                fx.Proxy(env[x.name], tracer) if isinstance(x, fx.Node) else x
+                for x in node.args
+            ]
+            output_proxy = decomposition_rules[node.target](*proxy_args)
+
+            # Operations on `Proxy` always yield new `Proxy`s, and the
+            # return value of our decomposition rule is no exception.
+            # We need to extract the underlying `Node` from the `Proxy`
+            # to use it in subsequent iterations of this transform.
+            new_node = output_proxy.node
+            env[node.name] = new_node
+        else:
+            # Default case: we don't have a decomposition rule for this
+            # node, so just copy the node over into the new graph.
+            new_node = new_graph.node_copy(node, lambda x: env[x.name])
+            env[node.name] = new_node
+    return fx.GraphModule(fx_module, new_graph)
+
+
+def tensorexpr_compile(fx_module, flat_args):
+    """Compiles the given fx_module using TensorExpr Kernel"""
+    inp_devices = set([i.device for i in flat_args if isinstance(i, torch.Tensor)])
+    assert len(inp_devices) == 1
+    inp_device = list(inp_devices)[0]
+    inputs = list()
+    output_refs = list()
+    for node in fx_module.graph.nodes:
+        if node.op == "placeholder":
+            inputs.append(node)
+        elif node.op == "output":
+            outputs = node.args[0]
+            if not isinstance(outputs, Iterable):
+                outputs = (outputs,)
+            new_outputs = []
+            for idx, output in enumerate(outputs):
+                # Appends (bool, idx) pairs
+                # if True, read from kernel outputs
+                # if False, read from kernel inputs
+                if output in inputs:
+                    output_refs.append((False, inputs.index(output)))
+                elif output in outputs[:idx]:
+                    output_refs.append((True, output_refs[outputs.index(output)][1]))
+                else:
+                    output_refs.append((True, len(new_outputs)))
+                    new_outputs.append(output)
+            node.args = (tuple(new_outputs),)
+    fx_module.graph.lint()
+    fx_module.recompile()
+
+    fx_module = decompose(fx_module)
+    for i in range(0, 100):
+        attr = f"_tensor_constant{i}"
+        if hasattr(fx_module, attr):
+            setattr(fx_module, attr, getattr(fx_module, attr).to(inp_device))
+        else:
+            break
+
+    jit_module = torch.jit.trace(fx_module, flat_args)
+    jit_module = torch.jit.freeze(jit_module.eval())
+    torch._C._jit_trace_module(jit_module._c, tuple(flat_args))
```

Comment:
we need to wait for Nick's PR to be merged before this.

Reply:
Thats ok. I can disable the testing of TEKernel for now and push this code in.